From 8845f6fbe89a8f8093d024b8270cc3d6f3e51bba Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:18:31 -0400
Subject: [fs] nfsd: nfs4_preprocess_seqid_op should only set *stpp on success

Message-id: <1410898849-22126-28-git-send-email-bfields@redhat.com>
Patchwork-id: 95803
O-Subject: [Patch RHEL7 027/165] nfsd: nfs4_preprocess_seqid_op should only set *stpp on success
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Not technically a bugfix, since nothing tries to use the return pointer
if this function doesn't return success, but it could be a problem
with some coming changes.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: Jeff Layton <jlayton@primarydata.com>
Reviewed-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: e17f99b728006fcebcf025b32fa7370bb998fb81
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 29788fd..71c442f 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -3953,6 +3953,7 @@ nfs4_preprocess_seqid_op(struct nfsd4_compound_state *cstate, u32 seqid,
 {
  __be32 status;
  struct nfs4_stid *s;
+ struct nfs4_ol_stateid *stp = NULL;
 
  dprintk("NFSD: %s: seqid=%d stateid = " STATEID_FMT "\n", __func__,
   seqid, STATEID_VAL(stateid));
@@ -3962,11 +3963,14 @@ nfs4_preprocess_seqid_op(struct nfsd4_compound_state *cstate, u32 seqid,
           cstate->minorversion, nn);
  if (status)
   return status;
- *stpp = openlockstateid(s);
+ stp = openlockstateid(s);
  if (!nfsd4_has_session(cstate))
-  cstate->replay_owner = (*stpp)->st_stateowner;
+  cstate->replay_owner = stp->st_stateowner;
 
- return nfs4_seqid_op_checks(cstate, stateid, seqid, *stpp);
+ status = nfs4_seqid_op_checks(cstate, stateid, seqid, stp);
+ if (!status)
+  *stpp = stp;
+ return status;
 }
 
 static __be32 nfs4_preprocess_confirmed_seqid_op(struct nfsd4_compound_state *cstate, u32 seqid,
-- 
1.7.1